### PR TITLE
AJ-1287 add e2e tests

### PR DESCRIPTION
**Pull request by calypsomatic**
[AJ-1111](https://broadworkbench.atlassian.net/browse/AJ-1111), [AJ-1287](https://broadworkbench.atlassian.net/browse/AJ-1287), [AJ-1139](https://broadworkbench.atlassian.net/browse/AJ-1139)

This PR adds:
- a python script that:
  - creates a workspace
  - verifies that WDS has started
  - uploads data to WDS
  - clones the workspace
  - verifies that the WDS data was cloned
 -  a github action that:
    - takes in a pre-existing bee with billing account and users
    - runs the python script
- resources to be used in the python script

The script has methods for starting a workflow that are not currently used; they can be adapted and turned on in the future for workflow-specific tests.  

Questions for reviewers:
- Should we run a single test that creates & clones the workspace, or should creating and cloning be separate tests?  If separate, would the cloning test take in a pre-existing workspace (and if so, where would it come from)?
- Is the logic that waits for WDS to start up correct?  Will it get stuck in an infinite loop?  Are there any cases that I'm missing? 

I welcome your thoughts on the organization of the files, how to modularize methods to be re-used in different tests, etc, although I expect this to evolve over time and many of those changes can come after this PR.

See https://github.com/DataBiosphere/terra-workspace-data-service/actions/runs/6227213226/job/16902800932 for a successful run of this action.

[AJ-1111]: https://broadworkbench.atlassian.net/browse/AJ-1111?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AJ-1287]: https://broadworkbench.atlassian.net/browse/AJ-1287?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AJ-1139]: https://broadworkbench.atlassian.net/browse/AJ-1139?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by jladieu** [comment]
> Should we run a single test that creates & clones the workspace, or should creating and cloning be separate tests? If separate, would the cloning test take in a pre-existing workspace (and if so, where would it come from)?

Might be more work, but I think having a separate/dedicated test for cloning would make the most sense, as that process could fail independently from creation and it'd give faster signal about cloning-specific issues.

I think having a standalone / persistent cloneable workspace that the clone script refers to would be fine for this, but the provisioning of that workspace should also itself be automated & rerunnable for easy teardown/rebuild.  Maybe a separate ticket?

File: .github/workflows/run-e2e-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  #AZURE_TOKEN: '${{ secrets.user-access-token }}' # github token for access to kick off a job in the private repo
+
+jobs:
+  init-github-context:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Get inputs or use defaults
+        id: extract-inputs
+        run: |
+#          echo "owner-subject=${{ inputs.owner-subject              || 'hermione.owner@quality.firecloud.org' }}" >> "$GITHUB_OUTPUT"
```
Comment (jladieu) [inline .github/workflows/run-e2e-tests.yaml:36]: I think this debug output would be helpful and not too spammy to leave enabled.

File: e2e-test/e2etest.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+
+from helper import *
+import os
+
+# Setup configuration
+# These values should be injected into the environment before setup
+azure_token = os.environ.get("AZURE_TOKEN")
+bee_name = os.environ.get("BEE_NAME")
+billing_project_name = os.environ.get("BILLING_PROJECT_NAME")
+#TODO: Take in arguments from the command line?
```
Comment (jladieu) [inline e2e-test/e2etest.py:10]: Agree with this TODO as it'd make it easier to do oneoff running, it'd make sense to do something like:

`commandLineInput || envVariableInput`

File: e2e-test/e2etest.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    setup(bee_name)
+    header = {"Authorization": "Bearer " + azure_token};
+
+    # create the workspace
+    workspace_id, workspace_name = create_workspace(cbas, billing_project_name, header)
+
+    # track to see when the workspace WDS is ready to upload data into them
+    # sleep for 5 minutes to allow WDS to start up, if no wds, only sleep 2 minutes to let cbas start up
+    if wds_upload:
+        time.sleep(200)
```
Comment (jladieu) [inline e2e-test/e2etest.py:27]: Might be worth printing something right before sleeping, and include the amount of time to be slept, so if you're actively watching this you can extrapolate when it will keep going.

**Comment by calypsomatic** [comment]
> Might be more work, but I think having a separate/dedicated test for cloning would make the most sense, as that process could fail independently from creation and it'd give faster signal about cloning-specific issues.
> 
> I think having a standalone / persistent cloneable workspace that the clone script refers to would be fine for this, but the provisioning of that workspace should also itself be automated & rerunnable for easy teardown/rebuild. Maybe a separate ticket?

We were talking about separating out the createWorkspace method to be another piece of the infrastructure (like creating the bee and billing project), for cases like this where it's not part of what's being tested, but yeah that will come later.

File: e2e-test/e2etest.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    elif cbas_submit_workflow:
+        time.sleep(120)
+    else:
+        print("TEST COMPLETE.")
+
+    if wds_upload:
```
Comment (jladieu) [inline e2e-test/e2etest.py:33]: Might be more complicated, but adjusting this logic to handle `provisioning` status and letting it poll for progress (and only fail when an error is detected) would make this test more resilient, and also potentially faster to proceed.  Instead of waiting 5 minutes minimum, you could poll every 30 seconds (like the UI does) and proceed ASAP on success.

File: .github/workflows/run-e2e-tests.yaml (hunk 1 above)
Comment (calypsomatic) [reply .github/workflows/run-e2e-tests.yaml:36]: This is not debug but actually generating a variable for use in a later step, except that we don't need it in this case.  (Other actions that might need different/many users or to differentiate between owners and readers)

File: e2e-test/e2etest.py (hunk 3 above)
Comment (calypsomatic) [reply e2e-test/e2etest.py:27]: Good call, I had a print statement about sleeping somewhere at some point in some iteration but seems like it got left out.

File: .github/workflows/run-e2e-tests.yaml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        uses: google-github-actions/auth@v1
+        with:
+          token_format: 'access_token'
+          workload_identity_provider: 'projects/1038484894585/locations/global/workloadIdentityPools/github-wi-pool/providers/github-wi-provider'
+          service_account: 'firecloud-qa@broad-dsde-qa.iam.gserviceaccount.com'
+          access_token_scopes: 'profile, email, openid'
+          access_token_subject: 'hermione.owner@quality.firecloud.org'
+          export_environment_variables: false 
+
+      - name: Run test
```
Comment (yuliadub) [inline .github/workflows/run-e2e-tests.yaml:72]: how owuld we know if the test fails? would we need to track the results of the github action manually or is that already hooked up somewhere? 

wondering if it makes sense to add this as an alert to a slack channel or something, if the test fails (or something similar)

File: e2e-test/e2etest.py (hunk 3 above)
Comment (yuliadub) [reply e2e-test/e2etest.py:27]: possible suggestion here to make this better - instead of sleeping actually pool for wds url/status in leo until you can get it (sometimes it may be faster then 200 seconds)

File: e2e-test/helper.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    logging.debug(f"rawls_url: {rawls_url}")
+    global leo_url
+    leo_url = f"https://leonardo.{bee_name}.bee.envs-terra.bio"
+    logging.debug(f"leo_url: {leo_url}")
+    return workspace_manager_url, rawls_url, leo_url
+
+# CREATE WORKSPACE ACTION
+def create_workspace(cbas, billing_project_name, header):
+    # create a new workspace, need to have attributes or api call doesnt work
+    api_call2 = f"{rawls_url}/api/workspaces";
```
Comment (yuliadub) [inline e2e-test/helper.py:33]: sorry this is totally my fault, but this can probably have a better name :D

Comment (yuliadub) [reply e2e-test/helper.py:33]: there are a few other ambiguous names in this file that can be improved, I think for clarify it would be useful to rename all the smthNumber to something more descriptive

**Comment by yuliadub** [comment]
thank you so much for including the cloning e2e test in this!!!

File: .github/workflows/run-e2e-tests.yaml (hunk 5 above)
Comment (calypsomatic) [reply .github/workflows/run-e2e-tests.yaml:72]: Ah good point I knew I was forgetting something.  I think the slack notifications would probably go in the calling workflow which would live in WDS.  I think if anything goes wrong it'll fail because something somewhere will throw an error but I could probably also put in some asserts.

File: e2e-test/e2etest.py (hunk 4 above)
Comment (calypsomatic) [reply e2e-test/e2etest.py:33]: So I am polling in the [get url method](https://github.com/broadinstitute/dsp-reusable-workflows/blob/abf375edc5e2efbe50212a0f866150723c5795d7/e2e-test/helper.py#L77).  I can reduce the initial sleep to some reasonable amount or remove it altogether.  Also definitely need some eyes on the polling logic to make sure it doesn't get stuck there forever

File: e2e-test/e2etest.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        logging.info(f"trying to see wds is ready to upload stuff for workspace {workspace_id}")
+        wds_url = get_app_url(workspace_id, "wds", azure_token)
```
Comment (jladieu) [inline e2e-test/e2etest.py:35]: Didn't realize while reading this first time through that `get_app_url` does polling (as evidenced by my PR comment =) -- maybe rename it to `poll_for_app_url` so that behavior is clear?

File: e2e-test/helper.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    request_body= {
+      "namespace": billing_project_name, # Billing project name
+      "name": f"api-workspace-{''.join(random.choices(string.ascii_lowercase, k=5))}", # workspace name
```
Comment (jladieu) [inline e2e-test/helper.py:36]: consider allowing the caller to pass in a a string which includes some disambiguating substring that will get included in the generated name (eg: 'wds-cloning-test') that way, if multiple tests start using this utility function, it'll be easier to tell where workspaces got created from for troubleshooting purposes

File: e2e-test/helper.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+      "attributes": {}};
+
+    response = requests.post(url=api_call2, json=request_body, headers=header)
+    logging.debug(f"url: {api_call2}")
+    logging.debug(f"headers: {header}")
+    logging.debug(f"response: {response}")
+    logging.debug(f"request_body: {request_body}")
+    
+    #example json that is returned by request: 'attributes': {}, 'authorizationDomain': [], 'bucketName': '', 'createdBy': 'yulialovesterra@gmail.com', 'createdDate': '2023-08-03T20:10:59.116Z', 'googleProject': '', 'isLocked': False, 'lastModified': '2023-08-03T20:10:59.116Z', 'name': 'api-workspace-1', 'namespace': 'yuliadub-test2', 'workspaceId': 'ac466322-2325-4f57-895d-fdd6c3f8c7ad', 'workspaceType': 'mc', 'workspaceVersion': 'v2'}
```
Comment (jladieu) [inline e2e-test/helper.py:45]: This is a helpful comment, but consider prettying up the formatting for easier consumption?

**Review by jladieu — APPROVED** [review]

File: e2e-test/e2etest.py (hunk 2 above)
Comment (calypsomatic) [reply e2e-test/e2etest.py:11]: I think I'm going to save this for future work, when other tests are being added and know what different parameters are needed.

File: e2e-test/helper.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+import logging
+
+logging.basicConfig(level=logging.DEBUG)
+
+workspace_manager_url = ""
+rawls_url = ""
+leo_url = ""
+
+def setup(bee_name):
+# define major service endpoints based on bee name
```
Comment (jladieu) [inline e2e-test/helper.py:18]: nit: indent

File: e2e-test/helper.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    logging.debug(f"leo_url: {leo_url}")
+    return workspace_manager_url, rawls_url, leo_url
+
+# CREATE WORKSPACE ACTION
+def create_workspace(cbas, billing_project_name, header, workspace_name = ""):
+    # create a new workspace, need to have attributes or api call doesnt work
+    rawls_workspace_api = f"{rawls_url}/api/workspaces";
+    request_body= {
+      "namespace": billing_project_name, # Billing project name
+      "name": workspace_name if workspace_name else f"api-workspace-{''.join(random.choices(string.ascii_lowercase, k=5))}",
```
Comment (jladieu) [inline e2e-test/helper.py:36]: Can you put the workspace name derivation logic into its own line and variable, and then log what actually gets used?  I think it may help troubleshooting.

File: e2e-test/helper.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    workspace_response_json = workpace_response.json()
+    logging.debug(f"json response: {workspace_response_json}")
+    data = json.loads(json.dumps(workspace_response_json))
+    
+    logging.debug(f"data['workspaceId']:  {data['workspaceId']}")
+    
+    # enable CBAS if specified
+    if cbas is True:
+        logging.info(f"Enabling CBAS for workspace {data['workspaceId']}")
+        start_cbas_api = f"{leo_url}/api/apps/v2/{data['workspaceId']}/terra-app-{str(uuid.uuid4())}";
```
Comment (jladieu) [inline e2e-test/helper.py:70]: nit: rename to `start_cbas_url`

Will requests.post log out the start_cbas_api?  If not, can you log the URL out?

**Review by jladieu — APPROVED** [review]

File: e2e-test/helper.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+    schema_client = wds_client.SchemaApi(api_client)
+
+    logging.info("verifying data was cloned")
+    response = schema_client.describe_record_type(workspaceId, version, recordName);
+    assert response.name == recordName, "Name does not match"
+    assert response.count == 2504, "Count does not match"
+
+
+def delete_workspace():
```
Comment (yuliadub) [inline e2e-test/helper.py:196]: since you are running this and then destroying the BEE - any reason to keep this in here? (its not a required to do for this per say)

**Comment by yuliadub** [comment]
> Questions for reviewers:
> 
> * Should we run a single test that creates & clones the workspace, or should creating and cloning be separate tests?  If separate, would the cloning test take in a pre-existing workspace (and if so, where would it come from)?
> * Is the logic that waits for WDS to start up correct?  Will it get stuck in an infinite loop?  Are there any cases that I'm missing?
> 
> I welcome your thoughts on the organization of the files, how to modularize methods to be re-used in different tests, etc, although I expect this to evolve over time and many of those changes can come after this PR.
> 
> See https://github.com/DataBiosphere/terra-workspace-data-service/actions/runs/6227213226/job/16902800932 for a successful run of this action.

- I think keeping the creation and cloning together works - no need to have a separate test since you need a source anyways and the first step of the test provides that for you
- left a comment in the PR- it is possible for it to get stuck in a forever true, but that is a small corner case. Easy way to resolve that would be to add a timer to the while loop. or alternatively you can just have the loop run through x times (like say since we sleep for an amount of time before running through it again, say running it 100 times is equivalent of a timer or something like that)


my only other comment would be around failures - today the script doesnt really protect against them (like an api timeout or something unrelated to wds failing specifically) could actually bring the test down. I dont think its urgent since you would have likely seen some of these failures in your testing if there were going to pop up, so this is something we can watch for as the test starts running on a schedule (and see if that is actually an issue)

**Review by yuliadub — APPROVED** [review]

File: e2e-test/helper.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+    api_client.configuration.host = wds_url
+    
+    # records client is used to interact with Records in the data table
+    records_client = wds_client.RecordsApi(api_client)
+    # data to upload to wds table
+    logging.info("uploading to wds")
+    # Upload entity to workspace data table with name "testType_uploaded"
+    response = records_client.upload_tsv(current_workspaceId, version, recordName, tsv_file_name)
+    logging.debug(response)
+    assert response.records_modified == 2504, f"Uploading to wds failed: {response.reason}"
```
[the remark this answers is not in the document]
Comment (calypsomatic) [reply e2e-test/helper.py:130]: Yeah that's in the test tsv that we're using for testing.  I'll see if I can pull it out

File: e2e-test/e2etest.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    header = {"Authorization": "Bearer " + azure_token};
+
+    # create the workspace
+    workspace_id, workspace_name = create_workspace(cbas, billing_project_name, header)
+
+    # track to see when the workspace WDS is ready to upload data into them
+    # sleep for 1 minute to allow apps to start up
+    if wds_upload or cbas_submit_workflow:
+        logging.info("Sleeping for 60 seconds while apps starts up")
+        time.sleep(60)
```
[the remark this answers is not in the document]
Comment (calypsomatic) [reply e2e-test/e2etest.py:27]: Since we know that it will take some amount of time for the apps to start up, shouldn't I at least start with some minimum?  I can put it down to 30 so it's equivalent to the polling loop going once
